File: lib/network/packets/server/s_breakthrough_result.dart
class SBreakthroughResult {
  const SBreakthroughResult({
    required this.success,
    required this.reason,
    required this.message,
    required this.realm,
    required this.realmStage,
    required this.realmLevel,
    required this.hp,
    required this.hpMax,
    required this.mp,
    required this.mpMax,
    required this.defense,
    required this.attack,
    required this.hit,
    required this.dodge,
    required this.hpRegen,
    required this.mpRegen,
    required this.puppetMax,
    required this.spellLearnRate,
    required this.craftProficiencyRate,
  });

  final bool success;
  final String reason;
  final String message;

  /// 以下欄位只有在 success 為 true 時才有意義
  final String realm;
  final int realmStage;
  final int realmLevel;
  final int hp;
  final int hpMax;
  final int mp;
  final int mpMax;
  final int defense;
  final int attack;
  final int hit;
  final int dodge;
  final int hpRegen;
  final int mpRegen;
  final int puppetMax;
  final int spellLearnRate;
  final int craftProficiencyRate;

  factory SBreakthroughResult.fromData(Map<String, dynamic> data) {
    return SBreakthroughResult(
      success: data['success'] as bool? ?? false,
      reason: data['reason'] as String? ?? '',
      message: data['message'] as String? ?? '',
      realm: data['realm'] as String? ?? '',
      realmStage: data['realmStage'] as int? ?? 0,
      realmLevel: data['realmLevel'] as int? ?? 0,
      hp: data['hp'] as int? ?? 0,
      hpMax: data['hpMax'] as int? ?? 0,
      mp: data['mp'] as int? ?? 0,
      mpMax: data['mpMax'] as int? ?? 0,
      defense: data['defense'] as int? ?? 0,
      attack: data['attack'] as int? ?? 0,
      hit: data['hit'] as int? ?? 0,
      dodge: data['dodge'] as int? ?? 0,
      hpRegen: data['hpRegen'] as int? ?? 0,
      mpRegen: data['mpRegen'] as int? ?? 0,
      puppetMax: data['puppetMax'] as int? ?? 0,
      spellLearnRate: data['spellLearnRate'] as int? ?? 100,
      craftProficiencyRate: data['craftProficiencyRate'] as int? ?? 100,
    );
  }
}

File: src/main/java/com/xin/server/packet/server/S_BreakthroughResult.java
package com.xin.server.packet.server;

import com.xin.server.model.instance.PcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;

/**
 * 境界突破結果回應。
 */
public class S_BreakthroughResult extends ServerBasePacket {

    public static final String REASON_OK = "OK";
    public static final String REASON_NOT_IN_GAME = "NOT_IN_GAME";
    public static final String REASON_MAX_REALM = "MAX_REALM";
    public static final String REASON_LEVEL_NOT_READY = "LEVEL_NOT_READY";
    public static final String REASON_MISSING_ITEM = "MISSING_ITEM";
    public static final String REASON_FAILED = "FAILED";

    private S_BreakthroughResult(boolean success, String reason, String message) {
        super(ServerOpcodes.S_BREAKTHROUGH_RESULT);
        put("success", success);
        put("reason", reason);
        put("message", message);
    }

    public static S_BreakthroughResult ok(PcInstance pc) {
        S_BreakthroughResult packet = new S_BreakthroughResult(true, REASON_OK, "突破成功");
        packet.put("realm", pc.getRealmName());
        packet.put("realmStage", pc.getRealmStage());
        packet.put("realmLevel", pc.getRealmLevel());
        packet.put("hp", pc.getCurrentHp());
        packet.put("hpMax", pc.getMaxHp());
        packet.put("mp", pc.getCurrentMp());
        packet.put("mpMax", pc.getMaxMp());
        packet.put("defense", pc.getDefense());
        packet.put("attack", pc.getAttack());
        packet.put("hit", pc.getHit());
        packet.put("dodge", pc.getDodge());
        packet.put("hpRegen", pc.getHpRegen());
        packet.put("mpRegen", pc.getMpRegen());
        packet.put("puppetMax", pc.getPuppetMax());
        packet.put("spellLearnRate", pc.getSpellLearnRate());
        packet.put("craftProficiencyRate", pc.getCraftProficiencyRate());
        return packet;
    }

    public static S_BreakthroughResult fail(String reason, String message) {
        return new S_BreakthroughResult(false, reason, message);
    }
}
